0015-ethereum-htlc-graceful-exit.adoc

```diff
@@ -0,0 +1,316 @@
+
+redeem:
+log1(0, 32, 0xB8CAC300E37F03AD332E581DEA21B2F0B84EAAADC184A295FEF71E81F44A7413) // log keccak256("Redeemed()")
+selfdestruct(<redeem_identity>)
+
+refund:
+log1(0, 0, 0x5D26862916391BF49478B2F5103B0720A842B45EF145A268F2CD1FB2AED55178) // log keccak256("Refunded()")
+selfdestruct(<refund_identity>)
+}
+----
+<1> If secret passed is the wrong size, it jumps to 4, does a `return`.
+<2> If the secret is incorrect, it does a `return`.
+<3> If the refund time is not expired, it does *not* jump and continues to 4 that does a `return`.
+
+As we can see, we currently execute a `return(0, 0)` for 3 reasons:
+
+A. The passed secret has a wrong length.
+B. The passed secret is incorrect (hashes do not match).
+C. No secret is passed but the time is not expired.
+
+In all cases, we currently pass `0,0` to the `return` instructions, meaning the returned output is `null` and all 3 cases are not differentiable by the caller.
+
+==== Using `return` instruction
+
+We could continue to use revert but use the returned output to provide more information on the execution of the contract.
+
+However, this would still mark a transaction that fails to redeem or refund as successful, hence this option is discarded.
+
+==== Using `revert` instruction
+
+Here is a propose update of the Ether Basic HTLC Atomic Swap using the `revert` instruction instead of `return`.
+Also, an error message is logged if the redeem or refund path fails.
+
+Note that while this example compiles with `solc`, it had not been tested.
+This should (hopefully) give an idea of how the HTLC would look like.
+
+For this example, we propose 2 error messages:
+- `wrongSecret`: if the secret is of incorrect length or it is not the correct secret (hash mismatch)
+- `tooEarly`: if no secret is provided and expiry has not yet happened
+
+.Proposed new Ether Basic HTLC using revert
+[source]
+----
+{
+    // Load received secret size
+    calldatasize
+
+    // Check if secret is zero length
+    iszero
+
+    // If secret is zero length, jump to branch that checks if expiry time has been reached
+    check_expiry
+    jumpi
+
+    // Load expected secret size
+    32
+
+    // Load received secret size
+    calldatasize
+
+    // Compare secret size
+    eq
+    iszero
+
+    // If passed secret is wrong size, jump to exit contract
+    exit_secret
+    jumpi
+
+    // Load secret into memory
+    calldatacopy(0, 0, 32)
+
+    // Hash secret with SHA-256 (pre-compiled contract 0x02)
+    call(72, 0x02, 0, 0, 32, 33, 32)
+
+    // Placeholder for correct secret hash
+    <secret_hash>
+
+    // Load hashed secret from memory
+    mload(33)
+
+    // Compare hashed secret with existing one
+    eq
+
+    // Combine `eq` result with `call` result
+    and
+
+    // Jump to redeem if hashes match
+    redeem
+    jumpi
+
+// Exit if it does not match
+exit_secret:
+    // Load Error Message "WRONG SECRET"
+    log1(0, 32, <keccak256("wrongSecret()")>)
+    // Exit
+    revert(0, 0)
```

Comment:
Since `revert` allows you to return something, did you consider something like this? 
`revert("tooEarly()")`

Reply:
I think @D4nte mentions somewhere that only _other_ contracts can read this return data, i.e. if you call the contract with a transaction, you don't have access to this. So it looks like we should emit a log to state what is going on!

Reply:
Excluded because only another contract using such function can access the returned data.
Returned data can be access from web3 if, and only if, it's a readonly function that can be `call`ed. See the `Return vs Log error message` section for more details.